### PR TITLE
Enabled C++ memory-map usage in PECOS-HNSW

### PR DESCRIPTION
**Pull request by OctoberChang**
*Issue #, if available:*

*Description of changes:*
- Enabled memory-map usage for C++ interface of PECOS-HNSW, which supports faster loading of the HNSW index.
- Fixed potential overview issues when offsetting memory pointer of PQ codebooks in HNSW-PQ4.

An C++ example code:
```
// train and save into memory-mapped format
pecos::ann::HNSW<float, feat_vec_t> indexer_;
indexer_.train(X_trn, M, efC, threads);
indexer_.save(output_dir);

// load from mmap-store, with lazy_load=false
pecos::ann::HNSW<float, feat_vec_t> indexer;
bool lazy_load = false;
indexer.load(output_dir , lazy_load);
``` 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pecos/core/ann/hnsw.hpp [hunk 1]
```diff
@@ -561,37 +514,37 @@ namespace ann {
...
+                std::string index_path = model_dir + "/index.mmap_store";
+                mmap_store.open(index_path.c_str(), lazy_load ? "r_lazy" : "r");
+                this->num_node = mmap_store.fget_one<index_type>();
+                this->maxM = mmap_store.fget_one<index_type>();
+                this->maxM0 = mmap_store.fget_one<index_type>();
+                this->efC = mmap_store.fget_one<index_type>();
+                this->max_level = mmap_store.fget_one<index_type>();
+                this->init_node = mmap_store.fget_one<index_type>();
+                graph_l0.load(mmap_store);
+                graph_l1.load(mmap_store);
```
Comment (rofuyu) [inline pecos/core/ann/hnsw.hpp:544]: let's add a comment here that we cannot call mmap_store.close() as the actual memory is held by this object.

File: pecos/core/ann/hnsw.hpp [hunk 2]
```diff
@@ -1160,43 +1117,43 @@ namespace ann {
...
+                this->maxM = mmap_store.fget_one<index_type>();
+                this->maxM0 = mmap_store.fget_one<index_type>();
+                this->efC = mmap_store.fget_one<index_type>();
+                this->max_level = mmap_store.fget_one<index_type>();
+                this->init_node = mmap_store.fget_one<index_type>();
+                this->subspace_dimension = mmap_store.fget_one<index_type>();
+                this->sub_sample_points = mmap_store.fget_one<index_type>();
+                feature_vec.load(mmap_store);
+                graph_l1.load(mmap_store);
+                graph_l0_pq4.load(mmap_store);
```
Comment (rofuyu) [inline pecos/core/ann/hnsw.hpp:1156]: add a similiar comment here.

File: test/tst-data/ann/hnsw-model-dense/c_model/config.json [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
         "init_node": 50,
         "maxM": 24,
         "maxM0": 48,
         "max_level": 1,
         "num_node": 90
-    }
-}
+    },
+    "version": "v2.0"
+}
```
Comment (rofuyu) [inline test/tst-data/ann/hnsw-model-dense/c_model/config.json]: add a newline here.

File: test/tst-data/ann/hnsw-model-dense/param.json [hunk 4]
```diff
@@ -15,4 +15,4 @@
   "topk": 10,
   "threads": 1
  }
-}
+}
```
Comment (rofuyu) [inline test/tst-data/ann/hnsw-model-dense/param.json]: add a newline here

File: test/tst-data/ann/hnsw-model-sparse/c_model/config.json [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
         "init_node": 50,
         "maxM": 24,
         "maxM0": 48,
         "max_level": 1,
         "num_node": 90
-    }
-}
+    },
+    "version": "v2.0"
+}
```
Comment (rofuyu) [inline test/tst-data/ann/hnsw-model-sparse/c_model/config.json]: add a new line here

File: test/tst-data/ann/hnsw-model-sparse/param.json [hunk 6]
```diff
@@ -15,4 +15,4 @@
   "topk": 10,
   "threads": 1
  }
-}
+}
```
Comment (rofuyu) [inline test/tst-data/ann/hnsw-model-sparse/param.json]: add a new line here.

**Review by rofuyu — COMMENTED** [review]
We need to fix all "Multiplication result converted to larger type" warning properly.

File: pecos/core/ann/hnsw.hpp (hunk 1 above)
Comment (OctoberChang) [reply pecos/core/ann/hnsw.hpp:544]: Fixed. See next revision.

File: pecos/core/ann/hnsw.hpp (hunk 2 above)
Comment (OctoberChang) [reply pecos/core/ann/hnsw.hpp:1156]: Fixed. See next revision.

File: test/tst-data/ann/hnsw-model-dense/c_model/config.json (hunk 3 above)
Comment (OctoberChang) [reply test/tst-data/ann/hnsw-model-dense/c_model/config.json]: Fixed. See next revision.

File: test/tst-data/ann/hnsw-model-dense/param.json (hunk 4 above)
Comment (OctoberChang) [reply test/tst-data/ann/hnsw-model-dense/param.json]: Fixed. See next revision.

File: test/tst-data/ann/hnsw-model-sparse/c_model/config.json (hunk 5 above)
Comment (OctoberChang) [reply test/tst-data/ann/hnsw-model-sparse/c_model/config.json]: Fixed. See next revision.

File: test/tst-data/ann/hnsw-model-sparse/param.json (hunk 6 above)
Comment (OctoberChang) [reply test/tst-data/ann/hnsw-model-sparse/param.json]: Fixed. See next revision.

**Review by rofuyu — APPROVED** [review]

**Pull request merged**